src/NLog/Layouts/Typed/Layout.cs

```diff
@@ -0,0 +1,374 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Globalization;
+using System.Text;
+using JetBrains.Annotations;
+using NLog.Common;
+using NLog.Config;
+using NLog.Internal;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout with a simple value (e.g. int) or a layout which results in a simple value (e.g. ${counter})
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class Layout<T> : Layout, IRenderable<T>
+    {
+        // ReSharper disable StaticMemberInGenericType - this is safe, static ctor is called for every generic type
+        [NotNull] private static readonly Type Type;
+        [NotNull] private static readonly string TypeNamed;
+        // ReSharper restore StaticMemberInGenericType
+
+        [CanBeNull] private readonly Layout _layout;
+        private readonly T _fixedValue;
+
+        /// <inheritdoc />
+        public Layout(T value)
+        {
+            _fixedValue = value;
+            IsFixed = true;
+            _layout = null;
+        }
+
+        /// <inheritdoc />
+        public Layout(Layout layout)
+        {
+            IsFixed = TryGetFixedValue(layout, out _fixedValue);
+            _layout = layout;
+        }
+
+        static Layout()
+        {
+            var type = typeof(T);
+            Type = Nullable.GetUnderlyingType(type) ?? type;
+            TypeNamed = type.Name;
+        }
+
+        /// <summary>
+        /// Is fixed value?
+        /// </summary>
+        public bool IsFixed { get; }
+
+        #region Overrides of TypedLayout<T>
+
+        private static string ValueToString(T value, CultureInfo cultureInfo)
+        {
+            if (value is IConvertible convertible)
+            {
+                return convertible.ToString(cultureInfo);
+            }
+
+            return value?.ToString();
+        }
+
+        private bool TryParse(string text, out T value, T defaultValue)
+        {
+            return TryConvertTo(text, out value, defaultValue);
+        }
+
+        private object _cacheKey = null;
+        private T _cacheValue = default(T);
+        private bool _cacheResult = false;
+
+        private bool TryCache(object raw)
+        {
+            return _cacheKey != null && (ReferenceEquals(_cacheKey, raw) || _cacheKey.Equals(raw));
```

Comment:
Maybe something like this:

```c#
        private bool TryGetCacheValue(object rawValue, out object resultValue, out bool resultValid)
        {
            var cacheKey = _cacheKey:
            var cacheValue = _cacheValue;
            var cacheValid = _cacheResult;
            if (!ReferenceEquals(cacheKey, null) && (ReferenceEquals(cacheKey, rawValue) || cacheKey.Equals(rawValue))
            {
                  resultValue = cacheValue;
                  resultValid = cacheValid;
                  return true;
            }
            resultValue = null;
            resultValid = false;
            return false;
       }
```

Reply:
is !ReferenceEquals(cacheKey, null) better than `cacheKey != null`?

Reply:
Think it compiles to the same. Just thought it would match the other `RereferenceEquals` :)

Reply:
I could also rename the current TryCache to IsCacheHit. Or do you really prefer the (longer) TryGetCacheValue?

Reply:
I like TryGet-method-pattern (returning bool and having out-parameter).

Reply:
applied